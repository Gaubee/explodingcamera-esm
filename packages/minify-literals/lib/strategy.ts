import CleanCSS from "clean-css";
import { type Options as HTMLOptions, minify } from "html-minifier-terser";
import type { TemplatePart } from "parse-literals";
import { randomBytes, randomInt } from "node:crypto";

/**
 * A strategy on how to minify HTML and optionally CSS.
 *
 * @template O minify HTML options
 * @template C minify CSS options
 */
export interface Strategy<O = any, C = any> {
	/**
	 * Retrieve a placeholder for the given array of template parts. The
	 * placeholder returned should be the same if the function is invoked with the
	 * same array of parts.
	 *
	 * The placeholder should be an HTML-compliant string that is not present in
	 * any of the parts' text.
	 *
	 * @param parts the parts to get a placeholder for
	 * @returns the placeholder
	 */
	getPlaceholder(parts: TemplatePart[], tag?: string): string | string[];
	/**
	 * Combines the parts' HTML text strings together into a single string using
	 * the provided placeholder. The placeholder indicates where a template
	 * expression occurs.
	 *
	 * @param parts the parts to combine
	 * @param placeholder the placeholder to use between parts
	 * @returns the combined parts' text strings
	 */
	combineHTMLStrings(parts: TemplatePart[], placeholder: string | string[]): string;
	/**
	 * Minfies the provided HTML string.
	 *
	 * @param html the html to minify
	 * @param options html minify options
	 * @returns minified HTML string
	 */
	minifyHTML(html: string, options?: O): Promise<string>;
	/**
	 * Minifies the provided CSS string.
	 *
	 * @param css the css to minfiy
	 * @param options css minify options
	 * @returns minified CSS string
	 */
	minifyCSS?(css: string, options?: C): Promise<string>;
	/**
	 * Splits a minfied HTML string back into an array of strings from the
	 * provided placeholder. The returned array of strings should be the same
	 * length as the template parts that were combined to make the HTML string.
	 *
	 * @param html the html string to split
	 * @param placeholder the placeholder to split by
	 * @returns an array of html strings
	 */
	splitHTMLByPlaceholder(html: string, placeholder: string | string[]): string[];
}

/**
 * The default <code>clean-css</code> options, optimized for production
 * minification.
 */
export const defaultMinifyCSSOptions: CleanCSS.Options = {};

/**
 * The default <code>html-minifier</code> options, optimized for production
 * minification.
 */
export const defaultMinifyOptions: HTMLOptions = {
	caseSensitive: true,
	collapseWhitespace: true,
	decodeEntities: true,
	minifyCSS: defaultMinifyCSSOptions,
	minifyJS: true,
	processConditionalComments: true,
	removeAttributeQuotes: false,
	removeComments: true,
	removeEmptyAttributes: true,
	removeScriptTypeAttributes: true,
	removeStyleLinkTypeAttributes: true,
	useShortDoctype: true,
};

/**
 * The default strategy. This uses <code>html-minifier</code> to minify HTML and
 * <code>clean-css</code> to minify CSS.
 */
export const defaultStrategy: Strategy<HTMLOptions, CleanCSS.Options> = {
	getPlaceholder(parts, tag) {
		const isCss = tag?.toLowerCase().includes("css");
		if (isCss) {
			// use strict mode to avoid issues with CSS minification
			const random = `tmp_${randomBytes(6).toString("hex")}`;
			const placeholder: string[] = [];
			const comment = /\/\*[\s\S]*?\*\//g;
			for (let i = 1; i < parts.length; i++) {
				const beforeFull = parts[i - 1]!.text;
				const beforeCss = beforeFull.replace(comment, "");
				const afterFull = parts[i]!.text;
				const afterCss = afterFull.replace(comment, "");
				/**
				 * 1. selector
				 * ${selector} {
				 * }
				 *
				 * 2. key
				 * selector {
				 * 	 ${key}: value;
				 * }
				 *
				 * 3. rule
				 * [selector {}]
				 * ${rule}
				 * [selector {}]
				 *
				 * 4. number-literal
				 * selector{
				 *   key: ${param}px;
				 * }
				 *
				 * 5. value
				 * selector {
				 * 	 key: ${value};
				 * 	 key: ${value}
				 * }
				 *
				 * 6. param
				 * selector{
				 *   key: fun(${param}[, ${param}]);
				 * }
				 */

				const isSelector = /^\s*\{/.test(afterCss);
				if (isSelector) {
					placeholder.push(`#${random}`);
					continue;
				}
				const isKey = /^\s*:/.test(afterCss);
				if (isKey) {
					placeholder.push(`--${random}`);
					continue;
				}
				const isRule = /\}\s*$/.test(beforeCss) || beforeCss.trim().length === 0;
				if (isRule) {
					return `@${random}();`;
				}
				const isUnit = /^\w+/.test(afterCss);
				if (isUnit) {
					let num: string;
					while (true) {
						num = `${randomInt(281474976710655)}`;
						if (!beforeFull.includes(num)) {
							break;
						}
					}
					placeholder.push(num);
					continue;
				}
				const isValue = /:\s*$/.test(beforeCss);
				if (isValue) {
					placeholder.push(`var(--${random})`);
					continue;
				}

				// isParams
				placeholder.push(`var(--${random})`);
			}
			return placeholder;
		}
		// Using @ and (); will cause the expression not to be removed in CSS.
		// However, sometimes the semicolon can be removed (ex: inline styles).
		// In those cases, we want to make sure that the HTML splitting also
		// accounts for the missing semicolon.
		const suffix = "();";
		let placeholder = "@TEMPLATE_EXPRESSION";
		while (parts.some((part) => part.text.includes(placeholder + suffix))) {
			placeholder += "_";
		}

		return placeholder + suffix;
	},

	combineHTMLStrings(parts, placeholder) {
		if (typeof placeholder === "string") {
			return parts.map((part) => part.text).join(placeholder);
		}
		return parts.map((part, i) => part.text + (placeholder[i] ?? "")).join("");
	},

	async minifyHTML(html, options = {}) {
		let minifyCSSOptions: HTMLOptions["minifyCSS"];

		if (html.match(/<!--(.*?)@TEMPLATE_EXPRESSION\(\);(.*?)-->/g)) {
			console.warn(
				"minify-literals: HTML minification is not supported for template expressions inside comments. Minification for this file will be skipped.",
			);
			return html;
		}

		html = html.replaceAll("<@TEMPLATE_EXPRESSION();", "<TEMPLATE_EXPRESSION___");
		html = html.replaceAll("</@TEMPLATE_EXPRESSION();", "</TEMPLATE_EXPRESSION___");

		if (options.minifyCSS) {
			if (options.minifyCSS !== true && typeof options.minifyCSS !== "function") {
				minifyCSSOptions = { ...options.minifyCSS };
			} else {
				minifyCSSOptions = {};
			}
		} else {
			minifyCSSOptions = false;
		}

		let adjustedMinifyCSSOptions: false | ReturnType<typeof adjustMinifyCSSOptions> = false;
		if (minifyCSSOptions) {
			adjustedMinifyCSSOptions = adjustMinifyCSSOptions(minifyCSSOptions);
		}

		let result = await minify(html, {
			...options,
			minifyCSS: adjustedMinifyCSSOptions,
		});

		result = result.replaceAll("<TEMPLATE_EXPRESSION___", "<@TEMPLATE_EXPRESSION();");
		result = result.replaceAll("</TEMPLATE_EXPRESSION___", "</@TEMPLATE_EXPRESSION();");

		if (options.collapseWhitespace) {
			// html-minifier does not support removing newlines inside <svg>
			// attributes. Support this, but be careful not to remove newlines from
			// supported areas (such as within <pre> and <textarea> tags).
			const matches = Array.from(result.matchAll(/<svg/g)).reverse();
			for (const match of matches) {
				const startTagIndex = match.index ?? 0;
				const closeTagIndex = result.indexOf("</svg", startTagIndex);
				if (closeTagIndex < 0) {
					// Malformed SVG without a closing tag
					continue;
				}

				const start = result.substring(0, startTagIndex);
				let svg = result.substring(startTagIndex, closeTagIndex);
				const end = result.substring(closeTagIndex);
				svg = svg.replace(/\r?\n/g, "");
				result = start + svg + end;
			}
		}
		result = fixCleanCssTidySelectors(html, result);

		return result;
	},
	async minifyCSS(css, options = {}) {
		const adjustedOptions = adjustMinifyCSSOptions(options);

		css = css.replaceAll(/@TEMPLATE_EXPRESSION\(\);:/g, "--TEMPLATE-EXPRESSION:");
		const output = await new CleanCSS({
			...adjustedOptions,
			returnPromise: true,
		}).minify(css);

		if (output.errors?.length) throw new Error(output.errors.join("\n\n"));

		// If there are warnings, return the unminified CSS.
		// CleanCSS can sometimes struggle with our preprocessed CSS due to the replaced template expressions.
		if (output.warnings?.length) console.log(css, output.styles);
		if (output.warnings.length) {
			console.warn(output.warnings.join("\n\n"));
			console.warn(
				"minify-literals: warnings during CSS minification, file was skipped. See above for details.",
			);
			return css.replace(/(\n)|(\r)/g, "");
		}

		output.styles = output.styles.replaceAll("--TEMPLATE-EXPRESSION:", "@TEMPLATE_EXPRESSION();:");
		output.styles = fixCleanCssTidySelectors(css, output.styles);
		return output.styles;
	},
	splitHTMLByPlaceholder(html, placeholder) {
		let parts: string[];
		if (typeof placeholder === "string") {
			parts = html.split(placeholder);
			// Make the last character (a semicolon) optional. See above.
			if (placeholder.endsWith(";")) {
				const withoutSemicolon = placeholder.substring(0, placeholder.length - 1);
				for (let i = parts.length - 1; i >= 0; i--) {
					parts.splice(i, 1, ...(parts[i]?.split(withoutSemicolon) ?? []));
				}
			}
		} else {
			parts = [];
			// strice mode
			let pos = 0;
			let index = -1;
			for (const ph of placeholder) {
				index = html.indexOf(ph, pos);
				if (index === -1) {
					throw new Error(`placeholder ${ph} not found in html ${html}`);
				}
				parts.push(html.slice(pos, index));
				pos = index + ph.length;
			}
			parts.push(html.slice(pos));
		}

		return parts;
	},
};

export function adjustMinifyCSSOptions(options: CleanCSS.Options = {}) {
	const level = options.level;

	const plugin = {
		level1: {
			value: (_name: any, value: string) => {
				if (!value.startsWith("@TEMPLATE_EXPRESSION") || value.endsWith(";")) return value;

				// The CSS minifier has removed the semicolon from the placeholder
				// and we need to add it back.
				return `${value};`;
			},
		},
	};

	return {
		...options,
		level,
		plugins: [plugin],
	};
}

// Should be fixed in clean-css https://github.com/clean-css/clean-css/issues/996, but is still happening
function fixCleanCssTidySelectors(original: string, result: string) {
	const regex = /(::?.+\((.*)\))[\s\r\n]*{/gm;
	let match: RegExpMatchArray | null;
	// biome-ignore lint/suspicious/noAssignInExpressions: this is fine
	while ((match = regex.exec(original)) != null) {
		const pseudoClass = match[1] ?? "";
		const parameters = match[2];

		if (!parameters?.match(/\s/)) {
			continue;
		}

		const parametersWithoutSpaces = parameters.replace(/\s/g, "");
		const resultPseudoClass = pseudoClass.replace(parameters, parametersWithoutSpaces);
		const resultStartIndex = result.indexOf(resultPseudoClass);
		if (resultStartIndex < 0) {
			continue;
		}

		const resultEndIndex = resultStartIndex + resultPseudoClass.length;
		// Restore the original pseudo class with spaces
		result = result.substring(0, resultStartIndex) + pseudoClass + result.substring(resultEndIndex);
	}

	return result;
}
