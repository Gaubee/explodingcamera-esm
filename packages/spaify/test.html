<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <title>Document</title>
  <script type="module">
    import { SPAify } from './dist/index.js';
    new SPAify({});
  </script>
</head>
<body>
  <main data-spaify-main>
    <h1>test</h1>
    <a href="./index.html">home</a>
  </main>
</body>
</html>