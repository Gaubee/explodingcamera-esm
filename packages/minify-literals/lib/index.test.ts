import { afterEach, beforeEach, describe, expect, it, test } from "bun:test";

import MagicString, { type SourceMapOptions } from "magic-string";
import { type ParseLiteralsOptions, type Template, type TemplatePart, parseLiterals } from "parse-literals";
import {
	type SourceMap,
	defaultGenerateSourceMap,
	defaultShouldMinify,
	defaultShouldMinifyCSS,
	defaultValidation,
	minifyHTMLLiterals,
} from "./";
import { defaultMinifyOptions, defaultStrategy } from "./strategy";

// https://github.com/explodingcamera/esm/issues/1
describe("handle key value pairs correctly", () => {
	it("should minify html", async () => {
		const source = `const css = css\`:host{\${"color"}: \${"red"}}\``;
		expect((await minifyHTMLLiterals(source))?.code).toMatch('const css = css`:host{${"color"}:${"red"}}`');
	});
});

// Comments https://github.com/asyncLiz/minify-html-literals/issues/49
describe("handle comments correctly", () => {
	it("should remove comments", async () => {
		const source = `const el = html\`<div><!-- comment --></div>\``;
		const expected = `const el = html\`<div></div>\``;
		expect((await minifyHTMLLiterals(source))?.code).toBe(expected);
	});

	it("templates inside of html comments are not minified", async () => {
		const source = `const el = html\`<div><!-- \${console.log(1)} --></div>\``;
		expect(await minifyHTMLLiterals(source)).toBe(null);
	});

	it("comments inside of literals are not removed", async () => {
		const source = `const el = html\`<div>\${/*console.log(1)*/}</div>\``;
		expect(await minifyHTMLLiterals(source)).toBe(null);
	});
});

// https://github.com/asyncLiz/minify-html-literals/issues/46
// currently, we skip the entire file if we find a template literal that uses unsafeCSS / unsafeHTML.
// instead, we should ideally just skip the template literal (but we need to update parse-literals for that)
describe("don't minify code with unsafeCSS / unsafeHTML", () => {
	it("should not remove styles from dynamically inserted selectors", async () => {
		const source = `
			css\`
				foo {
					bar: baz;
				}

				\${unsafeCSS('#foo-id')} {
					bar: baz;
				}
			\`
		`.trim();

		expect(await minifyHTMLLiterals(source)).toBe(null);
	});
});

// https://github.com/asyncLiz/minify-html-literals/issues/37
describe("minify templates with static tag literals", () => {
	it("should minify html", async () => {
		const STATIC_LITERAL_IN_TAG_NAME = `
			html\`<\${Component.litTagName} id="container">
					<span>Some content here</span>
				</\${Component.litTagName}>
				\`
		`.trim();

		const expected = `html\`<\${Component.litTagName} id="container"><span>Some content here</span></\${Component.litTagName}>\``;

		expect((await minifyHTMLLiterals(STATIC_LITERAL_IN_TAG_NAME))?.code).toBe(expected);
	});
});

test("example", async () => {
	const source = `
		const el = html\`<div > <h1>  Hello World  </h1 > </div>\`;
		const css = css\` .foo { color: red; }  \`;
	`;

	expect(await minifyHTMLLiterals(source)).toMatchSnapshot();
});

class MagicStringLike {
	generateMap(options?: Partial<SourceMapOptions>): SourceMap {
		return {
			version: 3,
			file: options?.file || null,
			sources: [options?.source || null],
			sourcesContent: [],
			names: [],
			mappings: "",
			toString() {
				return "";
			},
			toUrl() {
				return "";
			},
		};
	}

	overwrite(_start: number, _end: number, _content: string): any {
		// noop
	}

	toString(): string {
		return "";
	}
}

describe("minifyHTMLLiterals()", () => {
	const SOURCE = `
    function render(title, items, styles) {
      return html\`
        <style>
          \${styles}
        </style>
        <h1 class="heading">\${title}</h1>
        <button onclick="\${() => eventHandler()}"></button>
        <ul>
          \${items.map(item => {
            return getHTML()\`
              <li>\${item}</li>
            \`;
          })}
        </ul>
      \`;
    }

    function noMinify() {
      return \`
        <div>Not tagged html</div>
      \`;
    }

    function taggednoMinify(extra) {
      return other\`
        <style>
          .heading {
            font-size: 24px;
          }

          \${extra}
        </style>
      \`;
    }

    function taggedCSSMinify(extra) {
      return css\`
        .heading {
          font-size: 24px;
        }

        \${extra}
      \`;
    }

    function cssProperty(property) {
      const width = '20px';
      return css\`
        .foo {
          font-size: 1rem;
          width: \${width};
          color: \${property};
        }
      \`;
    }
  `;

	const SOURCE_MIN = `
    function render(title, items, styles) {
      return html\`<style>\${styles}</style><h1 class="heading">\${title}</h1><button onclick="\${() => eventHandler()}"></button><ul>\${items.map(item => {
            return getHTML()\`<li>\${item}</li>\`;
          })}</ul>\`;
    }

    function noMinify() {
      return \`
        <div>Not tagged html</div>
      \`;
    }

    function taggednoMinify(extra) {
      return other\`
        <style>
          .heading {
            font-size: 24px;
          }

          \${extra}
        </style>
      \`;
    }

    function taggedCSSMinify(extra) {
      return css\`.heading{font-size:24px}\${extra}\`;
    }

    function cssProperty(property) {
      const width = '20px';
      return css\`.foo{font-size:1rem;width:\${width};color:\${property}}\`;
    }
  `;

	const SVG_SOURCE = `
    function taggedSVGMinify() {
      return svg\`
        <svg xmlns="http://www.w3.org/2000/svg" height="24" viewBox="0 0 24 24" width="24">
          <path d="M6 19h12v2H6z" />
          <path d="M0 0h24v24H0V0z" fill="none" />
        </svg>
      \`;
    }
  `;

	const SVG_SOURCE_MIN = `
    function taggedSVGMinify() {
      return svg\`<svg xmlns="http://www.w3.org/2000/svg" height="24" viewBox="0 0 24 24" width="24"><path d="M6 19h12v2H6z"/><path d="M0 0h24v24H0V0z" fill="none"/></svg>\`;
    }
  `;

	const COMMENT_SOURCE = `
    function minifyWithComment() {
      return html\`
        <div .icon=\${0/*JS Comment */}>
        </div>
      \`;
    }
  `;

	const COMMENT_SOURCE_MIN = `
    function minifyWithComment() {
      return html\`<div .icon="\${0/*JS Comment */}"></div>\`;
    }
  `;

	const SVG_MULTILINE_SOURCE = `
    function multiline() {
      return html\`
        <pre>
          Keep newlines

          within certain tags
        </pre>
        <svg xmlns="http://www.w3.org/2000/svg" height="24" viewBox="0 0 24 24" width="24">
          <path d="M6 19h12v2H6z" />
          <path d="M0
                   0h24v24H0V0z"
                fill="none" />
        </svg>
      \`;
    }
  `;

	const SVG_MULTILINE_SOURCE_MIN = `
    function multiline() {
      return html\`<pre>
          Keep newlines

          within certain tags
        </pre><svg xmlns="http://www.w3.org/2000/svg" height="24" viewBox="0 0 24 24" width="24"><path d="M6 19h12v2H6z"/><path d="M0                   0h24v24H0V0z" fill="none"/></svg>\`;
    }
  `;

	const SHADOW_PARTS_SOURCE = `
    function parts() {
      return css\`
        foo-bar::part(.space .separated) {
          color: red;
        }
      \`;
    }
  `;

	const SHADOW_PARTS_SOURCE_MIN = `
    function parts() {
      return css\`foo-bar::part(.space .separated){color:red}\`;
    }
  `;

	const MEMBER_EXPRESSION_LITERAL_SOURCE = `
    function nested() {
      return LitHtml.html\`<div id="container">
        <span>Some content here</span>
      </div>
      \`;
    }
  `;

	const MEMBER_EXPRESSION_LITERAL_SOURCE_MIN = `
    function nested() {
      return LitHtml.html\`<div id="container"><span>Some content here</span></div>\`;
    }
  `;

	it('should minify "html" and "css" tagged templates', async () => {
		const result = await minifyHTMLLiterals(SOURCE, { fileName: "test.js" });

		expect(result).toBeTypeOf("object");
		expect(result!.code).toEqual(SOURCE_MIN);
	});

	it('should minify "svg" tagged templates', async () => {
		const result = await minifyHTMLLiterals(SVG_SOURCE, { fileName: "test.js" });
		expect(result).toBeTypeOf("object");
		expect(result!.code).toEqual(SVG_SOURCE_MIN);
	});

	it("should minify html with attribute placeholders that have no quotes and JS comments", async () => {
		const result = await minifyHTMLLiterals(COMMENT_SOURCE, { fileName: "test.js" });
		expect(result).toBeTypeOf("object");
		expect(result!.code).toEqual(COMMENT_SOURCE_MIN);
	});

	it("should minify html tagged with a member expression ending in html", async () => {
		const result = await minifyHTMLLiterals(MEMBER_EXPRESSION_LITERAL_SOURCE, {
			fileName: "test.js",
		});
		expect(result).toBeTypeOf("object");
		expect(result!.code).toEqual(MEMBER_EXPRESSION_LITERAL_SOURCE_MIN);
	});

	it("should minify multiline svg elements", async () => {
		const result = await minifyHTMLLiterals(SVG_MULTILINE_SOURCE, {
			fileName: "test.js",
		});
		expect(result).toBeTypeOf("object");
		expect(result!.code).toEqual(SVG_MULTILINE_SOURCE_MIN);
	});

	it("should not remove spaces in ::part()", async () => {
		const result = await minifyHTMLLiterals(SHADOW_PARTS_SOURCE, {
			fileName: "test.js",
		});
		expect(result).toBeTypeOf("object");
		expect(result!.code).toEqual(SHADOW_PARTS_SOURCE_MIN);
	});

	it("should return null if source is already minified", async () => {
		const result = await minifyHTMLLiterals(SOURCE_MIN, { fileName: "test.js" });
		expect(result).toBeNull();
	});

	it("should return a v3 source map", async () => {
		const result = await minifyHTMLLiterals(SOURCE, { fileName: "test.js" });
		expect(result).toBeTypeOf("object");
		expect(result!.map).toBeTypeOf("object");
		expect(result!.map!.version).toEqual(3);
		expect(result!.map!.mappings).toBeTypeOf("string");
	});

	describe("options", () => {
		it("should use defaultMinifyOptions", () => {
			minifyHTMLLiterals(SOURCE, { fileName: "test.js" });
			const parts = parseLiterals(SOURCE)[1]!.parts;
			defaultStrategy.combineHTMLStrings(parts, defaultStrategy.getPlaceholder(parts));
		});

		it("should allow custom partial minifyOptions", async () => {
			const minifyOptions = { caseSensitive: false };
			await minifyHTMLLiterals(SOURCE, { fileName: "test.js", minifyOptions });
			const parts = parseLiterals(SOURCE)[1]!.parts;
			defaultStrategy.combineHTMLStrings(parts, defaultStrategy.getPlaceholder(parts));
		});

		it("should use MagicString constructor", async () => {
			let msUsed: MagicStringLike | undefined;
			await minifyHTMLLiterals(SOURCE, {
				fileName: "test.js",
				generateSourceMap(ms) {
					msUsed = ms;
					return undefined;
				},
			});

			expect(msUsed).toBeInstanceOf(MagicString);
		});

		it("should allow custom MagicStringLike constructor", async () => {
			let msUsed: MagicStringLike | undefined;
			await minifyHTMLLiterals(SOURCE, {
				fileName: "test.js",
				MagicString: MagicStringLike,
				generateSourceMap(ms) {
					msUsed = ms;
					return undefined;
				},
			});

			expect(msUsed).toBeInstanceOf(MagicStringLike);
		});

		it("should allow custom parseLiterals()", async () => {
			const customParseLiterals = (source: string, options?: ParseLiteralsOptions) => {
				return parseLiterals(source, options);
			};

			await minifyHTMLLiterals(SOURCE, {
				fileName: "test.js",
				parseLiterals: customParseLiterals,
			});
		});

		it("should allow custom shouldMinify()", async () => {
			const customShouldMinify = (template: Template) => {
				return defaultShouldMinify(template);
			};

			await minifyHTMLLiterals(SOURCE, {
				fileName: "test.js",
				shouldMinify: customShouldMinify,
			});
		});

		it("should use defaultValidation", () => {
			minifyHTMLLiterals(SOURCE, {
				fileName: "test.js",
				strategy: {
					getPlaceholder: () => {
						return ""; // cause an error
					},
					combineHTMLStrings: defaultStrategy.combineHTMLStrings,
					minifyHTML: defaultStrategy.minifyHTML,
					splitHTMLByPlaceholder: defaultStrategy.splitHTMLByPlaceholder,
				},
			});

			expect(async () => {
				await minifyHTMLLiterals(SOURCE, {
					fileName: "test.js",
					strategy: {
						getPlaceholder: defaultStrategy.getPlaceholder,
						combineHTMLStrings: defaultStrategy.combineHTMLStrings,
						minifyHTML: defaultStrategy.minifyHTML,
						splitHTMLByPlaceholder: () => {
							return []; // cause an error
						},
					},
				});
			}).toThrow();
		});

		it("should allow disabling validation", async () => {
			await minifyHTMLLiterals(SOURCE, {
				fileName: "test.js",
				strategy: {
					getPlaceholder: () => {
						return ""; // cause an error
					},
					combineHTMLStrings: defaultStrategy.combineHTMLStrings,
					minifyHTML: defaultStrategy.minifyHTML,
					splitHTMLByPlaceholder: defaultStrategy.splitHTMLByPlaceholder,
				},
				validate: false,
			});
		});

		it("should allow disabling generateSourceMap", async () => {
			const result = await minifyHTMLLiterals(SOURCE, {
				fileName: "test.js",
				generateSourceMap: false,
			});
			expect(result).toBeTypeOf("object");
			expect(result!.map).toBeUndefined();
		});
	});

	describe("defaultShouldMinify()", () => {
		it('should return true if the template is tagged with any "html" text', () => {
			expect(defaultShouldMinify({ tag: "html", parts: [] })).toBe(true);
			expect(defaultShouldMinify({ tag: "HTML", parts: [] })).toBe(true);
			expect(defaultShouldMinify({ tag: "hTML", parts: [] })).toBe(true);
			expect(defaultShouldMinify({ tag: "getHTML()", parts: [] })).toBe(true);
			expect(defaultShouldMinify({ tag: "templateHtml()", parts: [] })).toBe(true);
		});

		it('should return false if the template is not tagged or does not contain "html"', () => {
			expect(defaultShouldMinify({ parts: [] })).toBe(false);
			expect(defaultShouldMinify({ tag: "css", parts: [] })).toBe(false);
		});

		it('should return true if the template is tagged with any "svg" text', () => {
			expect(defaultShouldMinify({ tag: "svg", parts: [] })).toBe(true);
			expect(defaultShouldMinify({ tag: "SVG", parts: [] })).toBe(true);
			expect(defaultShouldMinify({ tag: "sVg", parts: [] })).toBe(true);
			expect(defaultShouldMinify({ tag: "getSVG()", parts: [] })).toBe(true);
			expect(defaultShouldMinify({ tag: "templateSvg()", parts: [] })).toBe(true);
		});
	});

	describe("defaultShouldMinifyCSS()", () => {
		it('should return true if the template is tagged with any "css" text', () => {
			expect(defaultShouldMinifyCSS({ tag: "css", parts: [] })).toBe(true);
			expect(defaultShouldMinifyCSS({ tag: "CSS", parts: [] })).toBe(true);
			expect(defaultShouldMinifyCSS({ tag: "csS", parts: [] })).toBe(true);
			expect(defaultShouldMinifyCSS({ tag: "getCSS()", parts: [] })).toBe(true);
			expect(defaultShouldMinifyCSS({ tag: "templateCss()", parts: [] })).toBe(true);
		});

		it('should return false if the template is not tagged or does not contain "css"', () => {
			expect(defaultShouldMinifyCSS({ parts: [] })).toBe(false);
			expect(defaultShouldMinifyCSS({ tag: "html", parts: [] })).toBe(false);
		});
	});

	describe("defaultValidation", () => {
		describe("ensurePlaceholderValid()", () => {
			it("should throw an error if the placeholder is not a string", () => {
				expect(() => {
					defaultValidation.ensurePlaceholderValid(undefined);
				}).toThrow();
				expect(() => {
					defaultValidation.ensurePlaceholderValid(true);
				}).toThrow();
				expect(() => {
					defaultValidation.ensurePlaceholderValid({});
				}).toThrow();
			});

			it("should throw an error if the placeholder is an empty string", () => {
				expect(() => {
					defaultValidation.ensurePlaceholderValid("");
				}).toThrow();
			});

			it("should not throw an error if the placeholder is a non-empty string", () => {
				defaultValidation.ensurePlaceholderValid("EXP");
			});
		});
	});
});
