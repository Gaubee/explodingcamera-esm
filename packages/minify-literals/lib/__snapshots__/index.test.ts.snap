// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`example 1`] = `
{
  "code": 
"
		const el = html\`<div><h1>Hello World</h1></div>\`;
		const css = css\`.foo{color:red}\`;
	"
,
  "map": SourceMap {
    "file": ".map",
    "mappings": "AAAA;AACA,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,+BAAuC,CAAC,CAAC;AAC3D,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,eAAuB,CAAC,CAAC;AAC3C",
    "names": [],
    "sources": [
      ".map",
    ],
    "sourcesContent": undefined,
    "version": 3,
  },
}
`;
